community/kernel/src/main/java/org/neo4j/kernel/api/ExecutingQuery.java

```diff
@@ -111,17 +134,54 @@ public long startTime()
         return startTime;
     }
 
+    public long elapsedTime()
+    {
+        return clock.millis() - startTime;
+    }
+
+    /**
+     * @return the CPU time used by the query, in nanoseconds.
+     */
+    public long cpuTime()
+    {
+        return cpuClock.cpuTimeNanos( threadExecutingTheQuery ) - cpuTimeNanosWhenQueryStarted;
+    }
+
+    public long waitTime()
+    {
+        return TimeUnit.NANOSECONDS.toMillis( waitTimeNanos + status.waitTimeNanos( clock ) );
+    }
+
     @Override
     public String toString()
     {
-        return format(
-            "ExecutingQuery{queryId=%d, querySource='%s', username='%s', queryText='%s', queryParameters=%s, " +
-            "startTime=%d}",
-            queryId, querySource.toString( ":" ), username, queryText, queryParameters, startTime );
+        return ToStringBuilder.reflectionToString( this );
     }
 
     public Map<String,Object> metaData()
     {
         return metaData;
     }
+
+    public Locks.Tracer lockTracer()
+    {
+        return lockTracer;
+    }
+
+    public Map<String,Object> status()
+    {
+        return status.toMap( clock );
+    }
+
+    private Locks.WaitEvent waitForLock( ResourceType resourceType, long[] resourceIds )
+    {
+        ExecutingQueryStatus previous = status;
+        long startTimeNanos = clock.nanos();
+        status = new ExecutingQueryStatus.WaitingOnLock( resourceType, resourceIds, startTimeNanos );
+        return () ->
+        {
+            WAIT_TIME.addAndGet( ExecutingQuery.this, clock.nanos() - startTimeNanos );
+            status = previous;
+        };
```

Comment:
Do you have a concrete suggestion for how? Making the status object also implement `Locks.WaitEvent` feels like conflating concerns, but that is the only way I can think of reducing the number of allocations here.